lib/src/handle_access_token_response.dart

```diff
@@ -2,114 +2,86 @@
-
-    Map<String, dynamic> parameters;
-
-    if (contentType.mimeType == "application/x-www-form-urlencoded") {
-      parameters = {};
-
-      for (var unit in body.split('&')) {
-        var separator = unit.lastIndexOf('=');
-
-        // The '=' can't be the first or last character in a URL-encoded string
-        //
-        // For example, in 'a=b', the lowest index it can have is 1, and the greatest is
-        // `unit.length - 2`.
-        if (separator > 0 && separator < unit.length - 1) {
-          var key = unit.substring(0, separator);
-          var value = Uri.decodeComponent(unit.substring(separator + 1));
-          parameters[key] = value;
-        }
-      }
-    } else {
-      try {
-        var untypedParameters = JSON.decode(body);
-        validate(untypedParameters is Map,
-            'parameters must be a map, was "$parameters"');
-        parameters = DelegatingMap.typed(untypedParameters);
-      } on FormatException {
-        validate(false, 'invalid JSON');
-      }
-    }
-
-    return parameters;
-  };
-}
-
 /// Handles a response from the authorization server that contains an access
 /// token.
 ///
 /// This response format is common across several different components of the
 /// OAuth2 flow.
 /// 
 /// The scope strings will be separated by the provided [delimiter].
+///
+/// [getParameters] may be a function used to parse parameters out of responses from hosts that do not
+/// respond with application/json or application/x-www-form-urlencoded bodies. Use this
+/// to support authorization servers that do not completely follow the OAuth 2.0 standard.
 Credentials handleAccessTokenResponse(
     http.Response response,
     Uri tokenEndpoint,
     DateTime startTime,
     List<String> scopes,
     String delimiter,
-    {Map<String, dynamic> getParameters(String contentType, String body)}) {
-  if (response.statusCode != 200) _handleErrorResponse(response, tokenEndpoint);
+    {Map<String, dynamic> getParameters(String contentType, String body, Uri tokenEndpoint)}) {
+  getParameters ??= parseJsonOrUrlEncodedParameters;
 
-  validate(condition, message) =>
-      _validate(response, tokenEndpoint, condition, message);
-
-  getParameters ??= parseJsonOrUrlEncoded(validate);
+  if (response.statusCode != 200) _handleErrorResponse(response, tokenEndpoint, getParameters);
 
   var contentTypeString = response.headers['content-type'];
-  var parameters = getParameters(contentTypeString, response.body);
+  var parameters = getParameters(contentTypeString, response.body, tokenEndpoint);
 
   for (var requiredParameter in ['access_token', 'token_type']) {
-    validate(parameters.containsKey(requiredParameter),
-        'did not contain required parameter "$requiredParameter"');
-    validate(parameters[requiredParameter] is String,
-        'required parameter "$requiredParameter" was not a string, was '
-        '"${parameters[requiredParameter]}"');
+    try {
```

Comment:
Keep the old definition of `validate()`, and just wrap the call to `getParameters()` in a `try`/`catch` rather than wrapping each separate `validate()` call.

Reply:
👍